File: android/app/src/main/java/com/caliente/FaceLandmarkerModule.kt
package com.caliente

import android.graphics.BitmapFactory
import android.util.Base64
import com.facebook.react.bridge.*
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker


class FaceLandmarkerModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var faceLandmarker: FaceLandmarker? = null

    override fun getName(): String = "FaceLandmarkerModule"

    init {
        setupLandmarker(reactContext)
    }

    private fun setupLandmarker(context: ReactApplicationContext) {
        try {
            val baseOptions = BaseOptions.builder()
                .setModelAssetPath("face_landmarker.task")
                .build()

            val options = FaceLandmarker.FaceLandmarkerOptions.builder()
                .setBaseOptions(baseOptions)
                .setNumFaces(1)
                .setMinFaceDetectionConfidence(0.5f)
                .setMinFacePresenceConfidence(0.5f)
                .setMinTrackingConfidence(0.5f)
                .setOutputFaceBlendshapes(true)
                .setRunningMode(com.google.mediapipe.tasks.vision.core.RunningMode.IMAGE)
                .build()

            faceLandmarker = FaceLandmarker.createFromOptions(context, options)
            android.util.Log.d("MediaPipe", "FaceLandmarker initialized successfully")
        } catch (e: Exception) {
            android.util.Log.e("MediaPipe", "Failed to initialize: ${e.message}")
        }
    }

    @ReactMethod
    fun detectFace(imageData: String, promise: Promise) {
        try {
            val landmarker = faceLandmarker
            if (landmarker == null) {
                promise.reject("ERROR", "FaceLandmarker not initialized")
                return
            }

            val decodedBytes = Base64.decode(imageData, Base64.DEFAULT)
            val bitmap = BitmapFactory.decodeByteArray(decodedBytes, 0, decodedBytes.size)
            if (bitmap == null) {
                promise.reject("ERROR", "Invalid image data")
                return
            }

            val mpImage = BitmapImageBuilder(bitmap).build()
            val result = faceLandmarker!!.detect(mpImage)

            val response = Arguments.createMap()
            response.putBoolean("detected", false)
            response.putInt("landmarkCount", 0)

            if (result.faceLandmarks().isNotEmpty()) {
                val face = result.faceLandmarks()[0]
                response.putBoolean("detected", true)
                response.putInt("landmarkCount", face.size)

                val landmarks = Arguments.createArray()
                val keyIndices = intArrayOf(1, 4, 5, 6, 10, 33, 133, 152, 263, 362, 468, 473)
                for (index in keyIndices) {
                    if (index < face.size) {
                        val lm = face[index]
                        val point = Arguments.createMap()
                        point.putInt("index", index)
                        point.putDouble("x", lm.x().toDouble())
                        point.putDouble("y", lm.y().toDouble())
                        point.putDouble("z", lm.z().toDouble())
                        landmarks.pushMap(point)
                    }
                }
                response.putArray("landmarks", landmarks)
            }

            val blendshapeMap = Arguments.createMap()
            if (result.faceBlendshapes().isPresent && result.faceBlendshapes().get().isNotEmpty()) {
                val blendshapes = result.faceBlendshapes().get()[0]
                for (category in blendshapes) {
                    blendshapeMap.putDouble(
                        category.categoryName() ?: "unknown",
                        category.score().toDouble()
                    )
                }
            }
            response.putMap("blendshapes", blendshapeMap)

            promise.resolve(response)
        } catch (e: Exception) {
            promise.reject("ERROR", "Detection failed: ${e.message}")
        }
    }
}

File: ios/MediaPipe/FaceLandmarkerModule.swift
import Foundation
import MediaPipeTasksVision

@objc(FaceLandmarkerModule)
class FaceLandmarkerModule: NSObject {

  private var faceLandmarker: FaceLandmarker?

  override init() {
    super.init()
    setupLandmarker()
  }

  private func setupLandmarker() {
    guard let modelPath = Bundle.main.path(forResource: "face_landmarker", ofType: "task") else {
      print("[MediaPipe] Model file not found")
      return
    }

    do {
      let options = FaceLandmarkerOptions()
      options.baseOptions.modelAssetPath = modelPath
      options.numFaces = 1
      options.minFaceDetectionConfidence = 0.5
      options.minFacePresenceConfidence = 0.5
      options.minTrackingConfidence = 0.5
      options.outputFaceBlendshapes = true

      faceLandmarker = try FaceLandmarker(options: options)
      print("[MediaPipe] FaceLandmarker initialized successfully")
    } catch {
      print("[MediaPipe] Failed to initialize: \(error)")
    }
  }

  @objc func detectFace(_ imageData: String, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
    guard let landmarker = faceLandmarker else {
      reject("ERROR", "FaceLandmarker not initialized", nil)
      return
    }

    guard let data = Data(base64Encoded: imageData),
          let uiImage = UIImage(data: data),
          let mpImage = try? MPImage(uiImage: uiImage) else {
      reject("ERROR", "Invalid image data", nil)
      return
    }

    do {
      let result = try landmarker.detect(image: mpImage)

      var response: [String: Any] = [
        "detected": false,
        "landmarkCount": 0,
        "blendshapes": [:] as [String: Any]
      ]

      if let face = result.faceLandmarks.first {
        response["detected"] = true
        response["landmarkCount"] = face.count

        // Key landmarks
        var landmarks: [[String: Any]] = []
        let keyIndices = [1, 4, 5, 6, 10, 33, 133, 152, 263, 362, 468, 473]
        for index in keyIndices {
          if index < face.count {
            let lm = face[index]
            landmarks.append([
              "index": index,
              "x": lm.x,
              "y": lm.y,
              "z": lm.z
            ])
          }
        }
        response["landmarks"] = landmarks
      }

      if let blendshapes = result.faceBlendshapes.first {
        var blendshapeDict: [String: Any] = [:]
        for category in blendshapes.categories {
          blendshapeDict[category.categoryName ?? "unknown"] = category.score
        }
        response["blendshapes"] = blendshapeDict
      }

      resolve(response)
    } catch {
      reject("ERROR", "Detection failed: \(error.localizedDescription)", nil)
    }
  }

  @objc static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
